packages/@aws-cdk/aws-cognito/lib/user-pool-domain.ts

```diff
@@ -0,0 +1,129 @@
+import { ICertificate } from '@aws-cdk/aws-certificatemanager';
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { AwsCustomResource, AwsCustomResourcePolicy, AwsSdkCall, PhysicalResourceId } from '@aws-cdk/custom-resources';
+import { CfnUserPoolDomain } from './cognito.generated';
+import { IUserPool } from './user-pool';
+
+/**
+ * Represents a user pool domain.
+ */
+export interface IUserPoolDomain extends IResource {
+  /**
+   * The domain that was specified to be created.
+   * If `customDomain` was selected, this holds the full domain name that was specified.
+   * If the `cognitoDomain` was used, it contains the prefix to the Cognito hosted domain.
+   * @attribute
+   */
+  readonly domainName: string;
+}
+
+/**
+ * Options while specifying custom domain
+ * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html
+ */
+export interface CustomDomainOptions {
+  /**
+   * The custom domain name that you would like to associate with this User Pool.
+   */
+  readonly domainName: string;
+
+  /**
+   * The certificate to associate with this domain.
+   */
+  readonly certificate: ICertificate;
+}
+
+/**
+ * Options while specifying a cognito prefix domain.
+ * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-assign-domain-prefix.html
+ */
+export interface CognitoDomainOptions {
+  /**
+   * The prefix to the Cognito hosted domain name that will be associated with the user pool.
+   */
+  readonly domainPrefix: string;
+}
+
+/**
+ * Options to create a UserPoolDomain
+ */
+export interface UserPoolDomainOptions {
+  /**
+   * Associate a custom domain with your user pool
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html
+   * One of, and only one of, `customDomain` and `cognitoDomain` can be specified.
```

Comment:
If you're doing this, I'm wondering why it's not an enum-like...

Reply:
```ts
domain: Domain.cognitoDomain({ prefix: ... })

domain: Domain.customDomain({ domainName: ..., certificate: ... })
```

Reply:
I've modified it to be closer to your suggestion. How does that look?